Introduce a `search_throttled` threadpool

s1monw opened the pull request:
Today all searches happen on the search threadpool which is the correct
behavior in almost any case. Yet, there are exceptions where for instance
searches searches should be passed through a single-thread
threadpool to reduce impact on a node. This change adds a index-private setting that allows to mark an index as throttled for searches and forks off all non-stats searcher access to this threadpool for indices that are marked as `index.search.throttled`

elasticmachine commented:
Pinging @elastic/es-search-aggs

s1monw commented:
the majority of the change is cutting SearchService over to async execution. The actual addition of the setting and such is really a smallish change.

tlrx commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -405,59 +423,64 @@ private QueryFetchSearchResult executeFetchPhase(SearchContext context, SearchOp
-            cleanContext(context);
-        }
+    public void executeQueryPhase(InternalScrollSearchRequest request, SearchTask task,
+                                                     ActionListener<ScrollQuerySearchResult> listener) {
```
Nit: ident

tlrx commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1050,4 +1106,32 @@ public IndicesService getIndicesService() {
+            this.canMatch = canMatch;
+        }
+
+
```
Nit: extra line

jpountz commented:
I'm curious why we now fork can_match tasks unless the default threadpool is used, what is the benefit compared to never forking?

jasontedor commented on server/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -277,6 +278,26 @@
+            }
+            if (ThreadPool.Names.SEARCH.equals(s) == false && ThreadPool.Names.GENERIC.equals(s) == false &&
+                ThreadPool.THREAD_POOL_TYPES.containsKey(s)) {
+                throw new IllegalArgumentException("Invalid valid for [index.search.threadpool] - " + s + " is a reserved built-in " +
```
`valid` -> `value`

s1monw commented:
@jasontedor here is my suggestion. 
 1. I will add a `search_single` threadpool with a single thread and a queue.
 2. users can either decide to go to `search_single` or `search`
 3. in a seperate PR I add an option on the SearchRequest that does a _force put_ on the queue that way we can fix #33205 by special casing where necessary and not globally on the index.

@jpountz with 2. I think your question is more clear as well. for the  `search_single` case we need to make sure nobody accesses more than one reader of an index that is marked as `search_single` at the same time. 

I guess that makes this change more contained and addresses all concerns.

jasontedor commented:
@s1monw The plan sounds good, I think that we are mostly on the same page. I do wonder if we need to allow users to use `search_single` AKA `frozen`, i think we can do it automatically for frozen indices and otherwise use `search`? Then this would not need to be user configurable at all?

s1monw commented:
> @s1monw The plan sounds good, I think that we are mostly on the same page. I do wonder if we need to allow users to use search_single AKA frozen, i think we can do it automatically for frozen indices and otherwise use search? Then this would not need to be user configurable at all?

I can do that. Private setting sounds good to me. then this PR will only add the setting and the reading. The test can add a plugin that allows changing it.

jasontedor commented:
Thanks @s1monw.

jpountz commented:
The plan sounds good to me too. Then maybe call the setting "frozen_search" rather than "search_single" since it would be dedicated to frozen indices?

s1monw commented:
@jasontedor @jpountz I pushed more changes
